File: backend/tasks/taskCache.js
/**
 * Task cache module
 * Cache: task_id, dependencies (status for execution). Stage from backend. Db unchanged.
 */

const { computeTaskStages } = require('./taskStages');

const planCache = new Map(); // planId -> [{ task_id, dependencies }]

function extractCacheFromTask(task) {
  if (!task || !task.task_id) return null;
  return {
    task_id: task.task_id,
    dependencies: Array.isArray(task.dependencies) ? task.dependencies : []
  };
}

function extractCacheFromTasks(tasks) {
  if (!tasks || !Array.isArray(tasks)) return [];
  return tasks.map(t => ({
    task_id: t.task_id || '',
    dependencies: Array.isArray(t.dependencies) ? t.dependencies : [],
    status: t.status || 'undone'
  })).filter(c => c.task_id);
}

function getPlanCache(planId) {
  return planCache.get(planId) || [];
}

function appendPlanCache(planId, task) {
  const entry = extractCacheFromTask(task);
  if (!entry) return;
  const cache = getPlanCache(planId);
  const idx = cache.findIndex(c => c.task_id === entry.task_id);
  if (idx >= 0) cache[idx] = entry;
  else cache.push(entry);
  planCache.set(planId, cache);
}

function clearPlanCache(planId) {
  planCache.delete(planId);
}

function computeStaged(cacheEntries) {
  if (!cacheEntries || cacheEntries.length === 0) return [];
  return computeTaskStages(cacheEntries);
}

function enrichTreeData(staged, fullTasks) {
  const byId = new Map();
  (fullTasks || []).forEach(t => { if (t?.task_id) byId.set(t.task_id, t); });
  return staged.flat().map(c => {
    const full = byId.get(c.task_id) || {};
    return { ...full, ...c };  // 使用 c 的清洗后依赖，避免跨阶段连线
  });
}

/** Build flat treeData from tasks: extract cache -> compute stage -> enrich with full data. */
function buildTreeData(tasks) {
  if (!tasks || tasks.length === 0) return [];
  const cache = extractCacheFromTasks(tasks);
  const staged = computeStaged(cache);
  return enrichTreeData(staged, tasks);
}

module.exports = {
  extractCacheFromTask,
  extractCacheFromTasks,
  getPlanCache,
  appendPlanCache,
  clearPlanCache,
  computeStaged,
  enrichTreeData,
  buildTreeData
};

File: backend/tasks/task_cache.py
"""
Task pipeline: build_tree_data(tasks) for display/save.
Plan.tasks is the single source; pipeline output overwrites for render and persist.

Pipeline (each run, no old stage reused):
  1. Extract: task_id, dependencies (never stage)
  2. Sink: 依赖下沉
  3. Stage: 重新计算 stage (topological sort)
  4. Clean: 依赖清洗 (跨 stage 依赖，保险步骤)
  5. Enrich: 合并完整 task 信息
"""

from typing import List, Dict

from tasks.task_stages import compute_task_stages


def extract_cache_from_tasks(tasks: List[Dict]) -> List[Dict]:
    """Extract task_id, dependencies. Never stage (always recompute)."""
    if not tasks or not isinstance(tasks, list):
        return []
    return [
        {
            "task_id": t.get("task_id") or "",
            "dependencies": list(t.get("dependencies") or []),
        }
        for t in tasks
        if t.get("task_id")
    ]


def enrich_tree_data(staged: List[List[Dict]], full_tasks: List[Dict]) -> List[Dict]:
    by_id = {t["task_id"]: t for t in (full_tasks or []) if t.get("task_id")}
    result = []
    for stage in staged:
        for c in stage:
            full = by_id.get(c["task_id"], {})
            result.append({**full, **c})
    return result


def build_tree_data(tasks: List[Dict]) -> List[Dict]:
    """
    Build treeData from tasks. Never uses old stage.
    Pipeline: extract (no stage) -> sink -> recompute stage -> clean deps -> enrich.
    """
    if not tasks or len(tasks) == 0:
        return []
    cache = extract_cache_from_tasks(tasks)
    staged = compute_task_stages(cache)
    return enrich_tree_data(staged, tasks)
